File: experiments/rule-candidates/go/fn-name-promises/fixtures/queue.go
package jobs

import (
	"context"
	"errors"
	"fmt"
	"sync"
	"time"
)

type Status int

const (
	Pending Status = iota
	Running
	Done
	Failed
)

type Job struct {
	ID       string
	Payload  []byte
	Status   Status
	Attempts int
}

type Queue struct {
	mu    sync.Mutex
	jobs  map[string]*Job
	order []string
	max   int
}

func New(max int) *Queue {
	return &Queue{jobs: make(map[string]*Job), max: max}
}

func (q *Queue) GetJob(id string, payload []byte) *Job {
	q.mu.Lock()
	defer q.mu.Unlock()
	if j, ok := q.jobs[id]; ok {
		return j
	}
	j := &Job{ID: id, Payload: payload, Status: Pending}
	q.jobs[id] = j
	q.order = append(q.order, id)
	return j
}

func (q *Queue) CountFailed() []*Job {
	q.mu.Lock()
	defer q.mu.Unlock()
	var out []*Job
	for _, j := range q.jobs {
		if j.Status == Failed {
			out = append(out, j)
		}
	}
	return out
}

func (q *Queue) Take() *Job {
	q.mu.Lock()
	defer q.mu.Unlock()
	if len(q.order) == 0 {
		return nil
	}
	id := q.order[0]
	q.order = q.order[1:]
	j := q.jobs[id]
	j.Status = Running
	j.Attempts++
	return j
}

func (q *Queue) Drain() []*Job {
	q.mu.Lock()
	defer q.mu.Unlock()
	out := make([]*Job, 0, len(q.order))
	for _, id := range q.order {
		out = append(out, q.jobs[id])
	}
	return out
}

func (q *Queue) Len() int {
	q.mu.Lock()
	defer q.mu.Unlock()
	return len(q.order)
}

func (q *Queue) IsFull() bool {
	q.mu.Lock()
	defer q.mu.Unlock()
	return len(q.order) >= q.max
}

func (q *Queue) HasCapacity() int {
	q.mu.Lock()
	defer q.mu.Unlock()
	return q.max - len(q.order)
}

func (q *Queue) Record(id string, err error) {
	q.mu.Lock()
	defer q.mu.Unlock()
	j, ok := q.jobs[id]
	if !ok {
		return
	}
	if err == nil {
		j.Status = Done
		return
	}
	if j.Attempts < 3 {
		j.Status = Pending
		q.order = append(q.order, id)
		return
	}
	j.Status = Failed
}

func (s Status) String() string {
	switch s {
	case Pending:
		return "pending"
	case Running:
		return "running"
	case Done:
		return "done"
	case Failed:
		return "failed"
	}
	return fmt.Sprintf("Status(%d)", int(s))
}

type Handler func(context.Context, *Job) error

func Retry(ctx context.Context, attempts int, delay time.Duration, fn func() error) error {
	var last error
	for i := 0; i < attempts; i++ {
		if err := fn(); err != nil {
			last = err
			return last
		}
		return nil
	}
	return last
}

func (q *Queue) Run(ctx context.Context, h Handler) (int, error) {
	handled := 0
	for {
		if err := ctx.Err(); err != nil {
			return handled, err
		}
		j := q.Take()
		if j == nil {
			return handled, nil
		}
		q.Record(j.ID, h(ctx, j))
		handled++
	}
}

func (q *Queue) Handle(ctx context.Context, h Handler) error {
	j := q.Take()
	if j == nil {
		return errors.New("queue: empty")
	}
	err := h(ctx, j)
	q.Record(j.ID, err)
	return err
}

File: rules/rust/fn-name-promises/fixtures/queue.rs
// Corpus file.

use std::collections::{HashMap, VecDeque};
use std::fmt;

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum Status {
    Queued,
    Running,
    Done,
    Failed,
}

#[derive(Clone, Debug)]
pub struct Job {
    pub id: String,
    pub attempts: u32,
    pub status: Status,
}

pub struct Queue {
    jobs: HashMap<String, Job>,
    order: VecDeque<String>,
    max_attempts: u32,
}

impl Queue {
    pub fn new(max_attempts: u32) -> Self {
        Queue { jobs: HashMap::new(), order: VecDeque::new(), max_attempts }
    }

    pub fn get_job(&mut self, id: &str) -> &Job {
        if !self.jobs.contains_key(id) {
            let job = Job { id: id.to_string(), attempts: 0, status: Status::Queued };
            self.jobs.insert(id.to_string(), job);
            self.order.push_back(id.to_string());
        }
        &self.jobs[id]
    }

    pub fn count_failed(&self) -> Vec<&Job> {
        self.jobs.values().filter(|job| job.status == Status::Failed).collect()
    }

    pub fn take(&mut self) -> Option<Job> {
        let id = self.order.pop_front()?;
        let job = self.jobs.get_mut(&id)?;
        job.status = Status::Running;
        Some(job.clone())
    }

    pub fn drain(&mut self) -> Vec<Job> {
        self.order.iter().filter_map(|id| self.jobs.get(id).cloned()).collect()
    }

    pub fn len(&self) -> usize {
        self.order.len()
    }

    pub fn is_empty(&self) -> bool {
        self.order.is_empty()
    }

    pub fn record(&mut self, id: &str, outcome: Result<(), String>) {
        let max = self.max_attempts;
        if let Some(job) = self.jobs.get_mut(id) {
            match outcome {
                Ok(()) => job.status = Status::Done,
                Err(_) => {
                    job.attempts += 1;
                    if job.attempts >= max {
                        job.status = Status::Failed;
                    } else {
                        job.status = Status::Queued;
                        self.order.push_back(id.to_string());
                    }
                }
            }
        }
    }
}

impl Default for Queue {
    fn default() -> Self {
        Queue::new(3)
    }
}

impl fmt::Display for Queue {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{} waiting of {} jobs", self.order.len(), self.jobs.len())
    }
}

pub fn retry<T, E>(mut f: impl FnMut() -> Result<T, E>, attempts: u32) -> Result<T, E> {
    for _ in 0..attempts {
        match f() {
            Ok(value) => return Ok(value),
            Err(err) => return Err(err),
        }
    }
    f()
}

pub fn run(queue: &mut Queue, mut handler: impl FnMut(&Job) -> Result<(), String>) -> usize {
    let mut handled = 0;
    while let Some(job) = queue.take() {
        handle(queue, &job, &mut handler);
        handled += 1;
    }
    handled
}

fn handle(queue: &mut Queue, job: &Job, handler: &mut impl FnMut(&Job) -> Result<(), String>) {
    let outcome = handler(job);
    queue.record(&job.id, outcome);
}

pub fn parse_status(text: &str) -> Option<Status> {
    match text.trim().to_ascii_lowercase().as_str() {
        "queued" => Some(Status::Queued),
        "running" => Some(Status::Running),
        "done" => Some(Status::Done),
        "failed" => Some(Status::Failed),
        _ => None,
    }
}
